### PR TITLE
Expose toggle buttons in accessibility tree

### LINKED ISSUE #2444: Expose toggle buttons in accessibility tree
## Type of report

Feature request

## Provide description of the new feature

There are many toggle buttons inside editor's toolbar, e.g. "Bold", "Italic", "Underline" etc. To correctly expose them as toggle buttons in a11y tree, they must have `[role=button]` and `[aria-pressed]` attribute.

Default state:

```html
<a href="#" role="button" aria-pressed="false">Bold</a>
```

Toggled/pressed state:

```html
<a href="#" role="button" aria-pressed="true">Bold</a>
```

However we don't add `[aria-pressed=false]` for not-pressed buttons, as we do not distinguish between ordinary buttons and toggle ones. We only add `[aria-pressed=true]` when toggle buttons are activated. We could introduce the new toggle button category, which will always have `[aria-pressed]` attribute, removing handling of this attribute from ordinary buttons at the same time.

### LINKED ISSUE #4543: [Accessibility] - The selected state for the toolbar buttons is not announced by screen reader
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Launch ckeditor demo url https://ckeditor.com/ckeditor-4/demo/
2. Launch any screen reader like NVDA/JAWS/Voice Over(iOS).
3. Press bold and type some text.
4. Press ALT+F10 to navigate to toolbar using keyboard.
5. Navigate to Bold button and some other button which is not selected e.g. Italic.

### Expected result

For screen reader users proper information should be conveyed when they access toolbar buttons i.e. selected/not.

### Actual result

When screen reader users navigate to the selected item(Bold) and non selected item(Italic), screen reader is not conveying any information like Bold selected etc...

## Other details

* Browser: Chrome, Safari
* OS: Windows, iOS
* CKEditor version: 4

### PR DESCRIPTION
**Pull request by Comandeer**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [x] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
New features:

* [#2444](https://github.com/ckeditor/ckeditor4/issues/2444): Togglable toolbar buttons are now exposed as toggle buttons in the browser's accessibility tree.

Fixed Issues:

* [#4543](https://github.com/ckeditor/ckeditor4/issues/4543): The selected states of toolbar buttons are not announced by screen readers.
```

## What changes did you make?

I've changed the way of handling the `[aria-pressed]` attribute in the `button` plugin. After the changes, the attribute is always present on the button and only its value is changed (previously the attribute was removed when the button was not pressed). Additionally, I've added the `isToggle` option to the button's definition which indicates that the button is togglable (this option is needed to handle the initial state of the toggle buttons).

I've then added `isToogle` option to plugins using toggle buttons:

* basicstyles,
* bidi,
* blockquote,
* copyformatting,
* justify,
* list,
* sourcearea.

However, there are still some plugins that require additional attention:

* language and colorbutton – these plugins use menu buttons that are also togglable, however, screen readers do not support such combination, so the information about the state of the toggle button is lost,
* maximize – this plugin uses the toggle button but instead of conveying the state using the `[aria-pressed]` attribute, it changes the label of the button (from "Maximize" to "Minimize"; I'd suggest replacing the current method with the `[aria-pressed]` one, used in other plugins.

## Which issues does your PR resolve?

Closes #2444.
Closes #4543.

### CODE REVIEW

File: plugins/button/plugin.js
Comment (jacekbogdanski) [inline plugins/button/plugin.js:58]: I'm not sure the toggle template is a good candidate for moving it into the templates store. Do you see any use case where it would make sense to customize this template using the "template" event?

File: plugins/button/plugin.js
Comment (jacekbogdanski) [inline plugins/button/plugin.js:349]: I'm not sure if it's not introducing a breaking change. If someone is already using a button (3rd party plugin) without an arrow, it supports the `aria-pressed` feature. Maybe we should instead keep the previous logic for the non-defined `isToggle` value?

File: plugins/maximize/plugin.js
Comment (jacekbogdanski) [inline plugins/maximize/plugin.js:260]: Aren't we losing some information about the second behavior of the button with that change? Previously, this button was nicely labeled with `minimize` once maximized. Now, it looks like this change favors screen readers only. Is it possible to keep both an aria-pressed attribute and a label (in the context of accessibility)?

File: plugins/button/plugin.js
Comment (jacekbogdanski) [reply plugins/button/plugin.js:349]: Uh, alright, that wouldn't make much sense as we will be telling that minimalize is active.

**Review by jacekbogdanski — COMMENTED** [review]
TBC

File: plugins/button/plugin.js
Comment (Comandeer) [reply plugins/button/plugin.js:58]: I've not even know we have such an event :see\_no\_evil: I simply used a template here as all other button's parts were templates. But you're right, there's no need to have a template for the single attribute.

**Comment by Comandeer** [comment]
I've removed the unnecessary template for the `[aria-pressed]` attribute.

File: plugins/maximize/plugin.js
Comment (jacekbogdanski) [reply plugins/maximize/plugin.js:260]: Yeah, I think that makes sense, thanks for explanation 👍🏻

File: plugins/button/plugin.js
Comment (jacekbogdanski) [reply plugins/button/plugin.js:346]: Looks like it was setting `aria-pressed` for `!this.hasArrow` check? Although I don't see it to be working correctly before your changes on VoiceOver despite having `aria-pressed="true"` for toggled button - not sure why.

File: tests/plugins/button/_helpers/buttontools.js
Comment (jacekbogdanski) [inline tests/plugins/button/_helpers/buttontools.js:28]: Could you correct this if/else to use brackets?

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
These changes look good but I was only able to do manual tests using VoiceOver as my VM machines decided to not support OS sounds. @sculpt0r could you give it a try and confirm that manual test for screen readers works correctly on JAWS + Firefox and NVDA?

There is one issue but I'm not sure if it's related to your changes. Toggling source area forces the screen reader (VoiceOver) to read some garbage about being inside a table (for manual test with a table) or inside a list (for nightly sample). I'm not sure, but it looks like screen reader focus goes somewhere else after toggling a button?

**Comment by Comandeer** [comment]
> There is one issue but I'm not sure if it's related to your changes. Toggling source area forces the screen reader (VoiceOver) to read some garbage about being inside a table (for manual test with a table) or inside a list (for nightly sample). I'm not sure, but it looks like screen reader focus goes somewhere else after toggling a button?

That's weird :thinking: The focus does not go anywhere, it stays on the button. The issue you describe appears only in Chrome, I can't reproduce it in Firefox or Safari. It looks like a browser bug.

File: plugins/button/plugin.js
Comment (Comandeer) [reply plugins/button/plugin.js:346]: Because toggle buttons are identified by the presence of the `[aria-pressed]` attribute. If the attribute is not present, it's considered a normal button. So adding that attribute only for the `true` value causes that the button is a toggle one only when it's pressed (in Chrome it isn't at all).

**Comment by Comandeer** [comment]
I've fixed the reported codestyle issues.

File: tests/plugins/maximize/manual/togglebutton.md
Comment (sculpt0r) [inline tests/plugins/maximize/manual/togglebutton.md:6]:
```suggestion
1. Check the info about the "Maximize" button that is rendered inside the editor.
```

File: tests/plugins/button/_helpers/buttontools.js
Comment (sculpt0r) [inline tests/plugins/button/_helpers/buttontools.js:3]: Instead of the comment, we might rename the property  to:
```suggestion
	buttonAriaAttributes: {
```

or even something longer
```suggestion
	buttonStandardAriaAttributes: {
```

When I first read it - I wonder what are the `typical` attributes.

File: plugins/button/plugin.js
Comment (sculpt0r) [inline plugins/button/plugin.js:314]: Since it is part responsible only for arai attribute, then maybe:
```suggestion
				toggleAriaHtml: this.isToggle ? 'aria-pressed="false"' : ''
```

**Review by sculpt0r — CHANGES_REQUESTED** [review]
@Comandeer nice work with the test tooling 👍 

JAWS & NVDA works fine 🎉  I have only doubts about naming and other small stuff in a few places - please take a look at the inline comments.

Additionally - I think we should add a test with the language button in the toolbar. It is not the plain toggle button, but the code around `hasArrow` was changed and it will be good to verify whenever this button still works as we expected - also with screen readers.

File: plugins/button/plugin.js
Comment (Comandeer) [reply plugins/button/plugin.js:344]: Ok, I've rephrased the second comment, it should be clearer now.

**Comment by Comandeer** [comment]
Ok, I've refactored names a bit and added a manual test for screen readers and menu button.

**Review by sculpt0r — APPROVED** [review]
LGTM 🎉 

Thank you for the separated test for the menu button. Thanks to that, I found and extracted an issue with `aria-expanded`: #5144 

Added a changelog.

**Comment by sculpt0r** [comment]
@Comandeer - we have to wait for @jacekbogdanski approval - since he also reviewed this PR. Without it, GH is blocking me with the merge.

File: CHANGES.md
Comment (jacekbogdanski) [inline CHANGES.md:8]: A feature changelog entry should go to a major release.

**Review by jacekbogdanski — APPROVED** [review]

**Pull request merged**
